alf/utils/data_buffer.py

```diff
@@ -57,12 +62,15 @@ def _create_buffer(tensor_spec):
             dtype=tf.int32,
             shape=(),
             trainable=False)
-        # TF 2.0 checkpoint does not handle tuple. We have to convert
-        # _buffer to a flattened list in order to make the checkpointer save the
-        # content in self._buffer. This seems to be fixed in
-        # tf-nightly-gpu 2.1.0.dev20191119
-        # TODO: remove this after upgrading tensorflow
-        self._flattened_buffer = tf.nest.flatten(self._buffer)
+
+        with tf.device(self._device):
```

Comment:
put this before line 53

Reply:
The reason why I didn't include self._current_size and self._current_pos in tf.device is that when calling current_size(), clear() and pop(), they don't have to be transferred from one device to another. basically, only using device when the buffer content is accessed.

Reply:
At line 94, if indices is on GPU, but buf on CPU,  the system need to switch and transfer data between GPU and CPU, which  is slow.  And for the scalar like _current_size, they are mostly used  in CPU code. GPU can only do parallel computations like vector and matrix math.